libcontainer/nsenter/nsexec.c

```diff
@@ -664,40 +682,146 @@ void try_unshare(int flags, const char *msg)
+	pid_t pid;
+	int sig;
+
+	pid = getpid();
+
+	if (sigemptyset(&set)) {
 		sane_kill(pid, SIGKILL);
-		bail("failed to create userns path string");
+		bail("failed to empty sigset");
 	}
+	if (sigaddset(&set, SIGKILL)) {
+		sane_kill(pid, SIGKILL);
+		bail("failed to add SIGKILL to sigset");
+	}
+
+	if (sigwait(&set, &sig)) {
+		sane_kill(pid, SIGKILL);
+		bail("failed to sigwait");
+	}
+
+	return 0;
+}
+
+void send_idmapsources(int sockfd, pid_t pid, struct nlconfig_t *config)
+{
+	char *uidmap_src = config->idmap_uidmap;
+	char *gidmap_src = config->idmap_gidmap;
+
+	char *idmap_src = config->idmapsources;
+	char *idmap_end = idmap_src + config->idmapsources_len;
+
+	int nr_uidmap = 0;
+	int nr_gidmap = 0;
+	int nr_idmap = 0;
+
+	int i;
 
-	int userns_fd = open(proc_user_path, O_RDONLY | O_CLOEXEC | O_NOCTTY);
-	if (userns_fd < 0) {
+	/*
+	 * There should be as many UID/GID map as mount sources as done in
+	 * bootstrapData().
+	 * But let's check this nonethelss.
+	 */
+	for (i = 0; i < config->idmapsources_len; i++)
+		if (idmap_src[i] == '\0')
+			nr_idmap++;
+
+	for (i = 0; i < config->idmap_uidmap_len; i++)
+		if (uidmap_src[i] == '\0')
+			nr_uidmap++;
+
+	for (i = 0; i < config->idmap_gidmap_len; i++)
+		if (gidmap_src[i] == '\0')
+			nr_gidmap++;
+
+	if (nr_gidmap != nr_uidmap && nr_uidmap != nr_idmap) {
 		sane_kill(pid, SIGKILL);
-		bail("failed to get user namespace fd");
+		bail("there should be as many id map mount sources as UID and GID mappings: %d vs %d and %d", nr_idmap,
+		     nr_uidmap, nr_gidmap);
 	}
 
-	char *idmap_end = idmap_src + idmap_src_len;
 	while (idmap_src < idmap_end) {
+		char proc_user_path[PATH_MAX];
+		int child_stack_size = 4096;
+		void *child_stack;
+		int child_pid;
+		int ret;
+
 		if (idmap_src[0] == '\0') {
 			idmap_src++;
+			uidmap_src++;
+			gidmap_src++;
+
 			continue;
 		}
 
+		/*
+		 * For each mount source, we have corresponding UID and GID
+		 * mappings.
+		 * Then, we need to apply the man user_namespace for each of
+		 * them.
+		 * That is to say, create a new child with CLONE_NEWUSER, write
+		 * the corresponding UID and GID maps, then doing the ID map
+		 * mount with the child PID.
+		 * This could be parallelized by spawning a thread for each
+		 * triple (mount source, UID map, GID map).
+		 * But for the moment, let's keep it sequential.
+		 */
+		child_stack = malloc(child_stack_size);
+		if (child_stack == NULL) {
+			sane_kill(pid, SIGKILL);
+			bail("failed to allocate child stack");
+		}
+
+		child_pid = clone(child_mount_user_ns, child_stack + child_stack_size, CLONE_NEWUSER, NULL);
+		if (child_pid == -1) {
+			free(child_stack);
+			sane_kill(pid, SIGKILL);
+			bail("failed to clone with CLONE_NEWUSER");
+		}
+
+		update_setgroups(child_pid, SETGROUPS_DENY);
+
+		int uidmap_len = strlen(uidmap_src);
+		int gidmap_len = strlen(gidmap_src);
+
+		/* Update child mappings from the parent. */
+		update_uidmap(config->uidmappath, child_pid, uidmap_src, uidmap_len);
```

Comment:
Because we don't check the conflict of idmap items when `len(UIDMappings) > 0`, so will get an error in `update_uidmap`, but it uses `bail` to print the error and call `exit`, so it will not kill child_pid at that time and will got stuck. So we should make two changes:
1. validate there is a conflict of idmap items or not;
2. don't use bail to print error msg and return an error in `update_uidmap`. (https://github.com/opencontainers/runc/blob/a7afcd30cf85c4fa38bf1d181b519c42213223a5/libcontainer/nsenter/nsexec.c#L288-L301)

Reply:
I think there are the same problems in `update_setgroups` and `update_gidmap`.
So, when we clone a new process, we should double check whether there is a `bail` or not when we call a function.